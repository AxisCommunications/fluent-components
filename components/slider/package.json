{
  "name": "@axiscommunications/fluent-slider",
  "version": "8.6.0",
  "description": "Axis branded Slider",
  "homepage": "https://github.com/AxisCommunications/fluent-components#readme",
  "repository": {
    "type": "git",
    "url": "https://github.com/AxisCommunications/fluent-components.git",
    "directory": "components/slider"
  },
  "license": "MIT",
  "author": "Axis Communications AB",
  "main": "lib/index.js",
  "typings": "lib/index.d.ts",
  "files": [
    "lib"
  ],
  "scripts": {
    "build": "tsc",
    "check:unused-deps": "depcheck . --config=depcheck.yml",
    "lint": "tsc --noEmit && eslint . --cache",
    "test": "vitest run",
    "test:coverage": "vitest run --coverage",
    "test:watch": "vitest watch"
  },
  "devDependencies": {
    "@testing-library/jest-dom": "^6.1.4",
    "@testing-library/react": "^12.1.5",
    "@testing-library/user-event": "^14.5.1",
    "@types/react": "^17.0.67",
    "@types/react-dom": "^17.0.21",
    "@vitest/coverage-c8": "^0.33.0",
    "eslint": "^8.55.0",
    "jsdom": "^22.1.0",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "typescript": "^4.5.5",
    "vitest": "^0.34.6"
  },
  "peerDependencies": {
    "@fluentui/react-components": "^9.41.0",
    "@fluentui/react-icons": "^2.0.223",
    "@fluentui/react-utilities": "^9.15.2",
    "react": ">=16.8.0 <19.0.0",
    "react-dom": ">=16.8.0 <19.0.0",
    "scheduler": "^0.20.0"
  },
  "packageManager": "pnpm@8.1.0",
  "publishConfig": {
    "registry": "https://npm.pkg.github.com/"
  }
}
