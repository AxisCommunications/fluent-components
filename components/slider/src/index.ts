export type { MarkLabelProps } from "./mark/label/mark-label.types";
export type { MarkDef, MarkProps } from "./mark/mark.types";
export { RangeSlider } from "./range-slider";
export { Slider } from "./slider";
export type {
  RangeSliderOnChangeData,
  RangeSliderProps,
  SliderOnChangeData,
  SliderProps,
} from "./slider.types";
export type { ThumbProps } from "./thumb/thumb.types";
