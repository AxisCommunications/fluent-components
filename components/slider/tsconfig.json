{
  "extends": "./tsconfig.src.json",
  "compilerOptions": {
    "outDir": "lib",
    "lib": ["es5", "es6", "dom", "dom.iterable"]
  },
  "include": ["src"],
  "exclude": ["**/*.spec.*", "vitest.config.ts"]
}
