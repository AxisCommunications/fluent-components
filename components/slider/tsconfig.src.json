{
  "extends": "../../tsconfig.base.json",
  "include": ["src", "vitest.config.ts"],
  "compilerOptions": {
    "lib": ["es5", "es6", "dom", "dom.iterable"]
  }
}
