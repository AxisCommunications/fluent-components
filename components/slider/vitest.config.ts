// vite.config.ts
import { defineConfig } from "vitest/config";

export default defineConfig({
  test: {
    globals: true,
    environment: "jsdom",
    setupFiles: "./src/setupTest.ts",
    coverage: {
      provider: "v8",
      reporter: ["text", "html"],
      exclude: ["node_modules/"],
    },
  },
});
