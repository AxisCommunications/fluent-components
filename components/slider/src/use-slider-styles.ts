import {
  createFocusOutlineStyle,
  makeStyles,
  mergeClasses,
  shorthands,
  tokens,
  useFluent,
} from "@fluentui/react-components";

import { SliderState } from "./slider.types";

export const sliderClassNames = {
  root: "axis-Slider",
  control: "axis-Slider__control",
  rail: "axis-Slider__rail",
  track: "axis-Slider__track",
  thumb: {
    root: "axis-Slider__thumb",
    label: "axis-Slider__thumb-label",
  },
  mark: {
    root: "axis-Slider__mark",
    label: "axis-Slider__mark__label",
  },
};

export const sliderVars = {
  root: {
    paddingBottom: "--axis-slider--paddingBottom",
  },
  rail: {
    color: "--axis-slider__rail--color",
    size: "--axis-slider__rail--size",
  },
  track: {
    color: "--axis-slider__track--color",
  },
  thumb: {
    size: "--axis-Slider__thumb--size",
    color: "--axis-Slider__thumb--color",
  },
  tick: {
    color: "--axis-Slider__tick--color",
  },
  mark: {
    color: "--axis-Slider__mark--color",
  },
};

export const sliderDurations = {
  short: "100ms",
};

export const sliderEasings = {
  easeOutFast: "cubic-bezier(0.5, 0.5, 0, 1)",
};

const useRootStyles = makeStyles({
  root: {
    ...shorthands.padding(
      tokens.spacingVerticalSNudge,
      `0px`,
      `calc(var(${sliderVars.root.paddingBottom}) + ${tokens.spacingVerticalSNudge})`,
      `0px`
    ),
    height: `var(${sliderVars.thumb.size})`,
    display: "flex",
    position: "relative",
    width: `100%`,

    [sliderVars.root.paddingBottom]: "0px",
  },
  small: {
    [sliderVars.thumb.size]: "16px",
    [sliderVars.rail.size]: "2px",
  },
  medium: {
    [sliderVars.thumb.size]: "20px",
    [sliderVars.rail.size]: "4px",
  },
  focusIndicator: createFocusOutlineStyle({
    selector: "focus-within",
    style: {
      outlineOffset: {
        top: "-2px",
        bottom: "-2px",
        left: "8px",
        right: "8px",
      },
    },
  }),
  hasMarkLabel: {
    [sliderVars.root.paddingBottom]: "1rem",
  },
});

const useControlStyles = makeStyles({
  control: {
    position: "relative",
    display: "inline-block",
    height: "100%",
    marginLeft: "auto",
    marginRight: "auto",
    width: `calc(100% - var(${sliderVars.thumb.size}))`,
    cursor: "pointer",
  },
  enabled: {
    [sliderVars.rail.color]: tokens.colorNeutralStrokeAccessible,
  },
  disabled: {
    pointerEvents: "none",
    [sliderVars.rail.color]: tokens.colorNeutralBackgroundDisabled,
  },
});

const useRailStyles = makeStyles({
  root: {
    ...shorthands.borderRadius(tokens.borderRadiusXLarge),
    ...shorthands.overflow("hidden"),
    display: "block",
    position: "absolute",
    width: "100%",
    height: `var(${sliderVars.rail.size})`,
    top: "50%",
    transform: "translateY(-50%)",
    backgroundColor: `var(${sliderVars.rail.color})`,
  },
});

const useTrackStyles = makeStyles({
  root: {
    ...shorthands.borderRadius(tokens.borderRadiusXLarge),
    display: "block",
    position: "absolute",
    height: `var(${sliderVars.rail.size})`,
    top: "50%",
    transform: "translateY(-50%)",
    backgroundColor: `var(${sliderVars.track.color})`,
  },
  active: {
    [sliderVars.track.color]: tokens.colorCompoundBrandBackgroundPressed,
  },
  enabled: {
    [sliderVars.track.color]: tokens.colorCompoundBrandBackground,
  },
  disabled: {
    [sliderVars.track.color]: tokens.colorNeutralForegroundDisabled,
  },
});

export const useSliderStyles_unstable = (state: SliderState): SliderState => {
  const rootStyles = useRootStyles();
  const controlStyles = useControlStyles();
  const railStyles = useRailStyles();
  const trackStyles = useTrackStyles();

  const { disabled, trackOffset, trackWidth, active, markLabels, size } = state;

  const { dir } = useFluent();

  const rtl = dir === "rtl";

  const hasMarkLabel = markLabels.length > 0;

  state.root.className = mergeClasses(
    sliderClassNames.root,
    rootStyles.root,
    size === "small" && rootStyles.small,
    size === "medium" && rootStyles.medium,
    rootStyles.focusIndicator,
    hasMarkLabel && rootStyles.hasMarkLabel,
    state.root.className
  );

  state.control.className = mergeClasses(
    sliderClassNames.control,
    controlStyles.control,
    disabled ? controlStyles.disabled : controlStyles.enabled,
    state.control.className
  );

  state.rail.className = mergeClasses(
    sliderClassNames.rail,
    railStyles.root,
    state.rail.className
  );

  state.track.className = mergeClasses(
    sliderClassNames.track,
    trackStyles.root,
    disabled ? trackStyles.disabled : trackStyles.enabled,
    active && trackStyles.active,
    state.track.className
  );

  const offsetDirection = rtl ? "right" : "left";
  state.track.style = {
    [offsetDirection]: `${trackOffset}%`,
    width: `${trackWidth}%`,
    ...state.track.style,
  };

  return state;
};
