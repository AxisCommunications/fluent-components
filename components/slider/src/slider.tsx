import type { ForwardRefComponent } from "@fluentui/react-utilities";
import React from "react";

import { renderSlider_unstable } from "./render-slider";
import type { SliderProps } from "./slider.types";
import { useSlider_unstable } from "./use-slider";
import { useSliderContextValues } from "./use-slider-context-values";
import { useSliderStyles_unstable } from "./use-slider-styles";

export const Slider: ForwardRefComponent<SliderProps> = React.forwardRef(
  (props, ref) => {
    const state = useSlider_unstable(props, ref);
    const contextValues = useSliderContextValues(state);

    useSliderStyles_unstable(state);

    return renderSlider_unstable(state, contextValues);
  }
);

Slider.displayName = "Slider";
