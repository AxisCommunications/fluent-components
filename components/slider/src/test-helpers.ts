import { vi } from "vitest";
import { sliderClassNames } from "./use-slider-styles";

export const getControlRoot = (sliderRoot: HTMLElement): HTMLElement => {
  const element = sliderRoot.querySelector(`.${sliderClassNames.control}`);
  if (element === null) {
    throw new Error("Unable to find control element");
  }
  element.getBoundingClientRect = () => ({
    x: 0,
    y: 0,
    width: 100,
    height: 10,
    bottom: 10,
    left: 0,
    right: 0,
    top: 0,
    toJSON: vi.fn(),
  });
  return element as HTMLElement;
};
