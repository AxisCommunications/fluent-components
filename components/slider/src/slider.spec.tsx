import { fireEvent, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";
import { FluentProvider } from "@fluentui/react-components";
import React from "react";
import { describe, expect, it, vi } from "vitest";
import { Slider } from "./slider";
import { getControlRoot } from "./test-helpers";
import { sliderClassNames } from "./use-slider-styles";

const expectSliderValue = (element: HTMLElement, value: number) => {
  expect(element.getAttribute("value")).toEqual(value.toString());
};

describe("slider", () => {
  it("should render", () => {
    const result = render(<Slider value={20} marks={[]} min={0} max={100} />);

    expect(result).toMatchSnapshot();
  });

  it("should apply value", () => {
    const { getByRole } = render(<Slider min={0} max={100} value={10} />);
    expect(getByRole("slider").getAttribute("value")).toEqual("10");
  });

  it("should apply default value", () => {
    const { getByRole } = render(
      <Slider min={0} max={100} defaultValue={10} />
    );
    expect(getByRole("slider").getAttribute("value")).toEqual("10");
  });

  it("should render marks", () => {
    const { getAllByTestId } = render(
      <Slider
        min={0}
        max={100}
        mark={<span data-testid="mark" />}
        marks={[{ value: 10 }, { value: 20 }]}
      />
    );
    expect(getAllByTestId("mark")).toHaveLength(2);
  });

  it("should render mark labels", () => {
    const { getByTestId } = render(
      <Slider
        min={0}
        max={100}
        marks={[
          { value: 10, label: "10" },
          { value: 20, label: "20" },
        ]}
        data-testid="slider-root"
      />
    );
    const root = getByTestId("slider-root");
    expect(
      root.querySelectorAll(`.${sliderClassNames.mark.label}`)
    ).toHaveLength(2);
  });

  it("should generate marks from step", () => {
    const { getAllByTestId } = render(
      <Slider
        min={0}
        max={100}
        mark={<span data-testid="mark" />}
        step={10}
        marks={true}
      />
    );
    expect(getAllByTestId("mark")).toHaveLength(11);
  });

  it("should transform value", () => {
    const { getByTestId, getByRole } = render(
      <Slider
        min={0}
        max={100}
        defaultValue={50}
        valueLabelTransform={(value) => value * 2}
        data-testid="slider"
      />
    );
    const sliderRoot = getByTestId("slider");
    const label = sliderRoot.querySelector(`.${sliderClassNames.thumb.label}`);
    expect(label).toHaveTextContent("100");
    expectSliderValue(getByRole("slider"), 50);
  });

  describe("mouse interaction", () => {
    it("should set value on click", () => {
      const { getByTestId, getByRole } = render(
        <Slider min={0} max={100} data-testid="slider-root" />
      );

      const controlRoot = getControlRoot(getByTestId("slider-root"));

      fireEvent.mouseDown(controlRoot, { button: 0, clientX: 75 });

      expect(getByRole("slider").getAttribute("value")).toEqual("75");
    });

    it("should set value on click when rtl", () => {
      const { getByTestId, getByRole } = render(
        <FluentProvider dir="rtl" theme={{}}>
          <Slider min={0} max={100} data-testid="slider-root" />
        </FluentProvider>
      );

      const controlRoot = getControlRoot(getByTestId("slider-root"));

      fireEvent.mouseDown(controlRoot, { button: 0, clientX: 75 });

      expect(getByRole("slider").getAttribute("value")).toEqual("25");
    });

    it("should focus thumb", () => {
      const { getByTestId, getByRole } = render(
        <Slider min={0} max={100} data-testid="slider-root" />
      );

      const slider = getByRole("slider");
      expect(document.activeElement).not.toEqual(slider);

      const controlRoot = getControlRoot(getByTestId("slider-root"));
      fireEvent.mouseDown(controlRoot, { button: 0, clientX: 0 });

      expect(document.activeElement).toEqual(slider);
    });

    it("should be disabled", () => {
      const { getByTestId } = render(
        <Slider min={0} max={100} data-testid="slider-root" disabled />
      );

      expect(getByTestId("slider-root").querySelector("input")).toBeDisabled();
    });

    it("should ignore non-left click", () => {
      const { getByTestId, getByRole } = render(
        <Slider min={0} max={100} data-testid="slider-root" />
      );

      const controlRoot = getControlRoot(getByTestId("slider-root"));

      fireEvent.mouseDown(controlRoot, { button: 1, clientX: 50 });
      expectSliderValue(getByRole("slider"), 0);
    });

    it("should call onChange but not committed on down", () => {
      const onChange = vi.fn();
      const onChangeCommitted = vi.fn();
      const { getByTestId } = render(
        <Slider
          min={0}
          max={100}
          onChange={onChange}
          onChangeCommitted={onChangeCommitted}
          data-testid="slider-root"
        />
      );

      const controlRoot = getControlRoot(getByTestId("slider-root"));
      fireEvent.mouseDown(controlRoot, { button: 0, clientX: 50 });

      expect(onChange).toHaveBeenCalledWith({ value: 50 });
      expect(onChangeCommitted).not.toHaveBeenCalled();
    });

    it("should call onChange but not committed on move", () => {
      const onChange = vi.fn();
      const onChangeCommitted = vi.fn();
      const { getByTestId } = render(
        <Slider
          min={0}
          max={100}
          onChange={onChange}
          onChangeCommitted={onChangeCommitted}
          data-testid="slider-root"
        />
      );

      const controlRoot = getControlRoot(getByTestId("slider-root"));
      fireEvent.mouseDown(controlRoot, { button: 0, clientX: 50 });
      fireEvent.mouseMove(controlRoot, { buttons: 1, clientX: 70 });

      expect(onChange).toHaveBeenCalledWith({ value: 70 });
      expect(onChangeCommitted).not.toHaveBeenCalled();
    });

    it("should call onChangeCommitted", () => {
      const onChangeCommitted = vi.fn();
      const { getByTestId } = render(
        <Slider
          min={0}
          max={100}
          onChangeCommitted={onChangeCommitted}
          data-testid="slider-root"
        />
      );

      const controlRoot = getControlRoot(getByTestId("slider-root"));
      fireEvent.mouseDown(controlRoot, { button: 0, clientX: 50 });
      fireEvent.mouseUp(controlRoot, { button: 0, clientX: 50 });

      expect(onChangeCommitted).toHaveBeenCalledWith({ value: 50 });
    });

    it("should move thumb to value", () => {
      const { getByTestId, getByRole } = render(
        <Slider min={0} max={100} defaultValue={10} data-testid="slider-root" />
      );

      const controlRoot = getControlRoot(getByTestId("slider-root"));
      fireEvent.mouseDown(controlRoot, { button: 0, clientX: 10 });
      fireEvent.mouseMove(controlRoot, { buttons: 1, clientX: 30 });
      fireEvent.mouseUp(controlRoot, { button: 0, clientX: 30 });

      expectSliderValue(getByRole("slider"), 30);
    });

    it("should abort thumb movement if button is no longer pressed", () => {
      const { getByTestId, getByRole } = render(
        <Slider min={0} max={100} defaultValue={10} data-testid="slider-root" />
      );

      const controlRoot = getControlRoot(getByTestId("slider-root"));
      fireEvent.mouseDown(controlRoot, { button: 0, clientX: 10 });
      fireEvent.mouseMove(controlRoot, { buttons: 0, clientX: 30 });

      expectSliderValue(getByRole("slider"), 10);
    });

    it("should snap to value", () => {
      const { getByTestId, getByRole } = render(
        <Slider min={0} max={100} step={20} data-testid="slider-root" />
      );

      const controlRoot = getControlRoot(getByTestId("slider-root"));
      fireEvent.mouseDown(controlRoot, { button: 0, clientX: 16 });

      expectSliderValue(getByRole("slider"), 20);
    });

    it("should snap to marks", () => {
      const { getByTestId, getByRole } = render(
        <Slider
          min={0}
          max={100}
          marks={[{ value: 20 }, { value: 60 }]}
          step="marks"
          data-testid="slider-root"
        />
      );

      const controlRoot = getControlRoot(getByTestId("slider-root"));
      fireEvent.mouseDown(controlRoot, { button: 0, clientX: 50 });

      expectSliderValue(getByRole("slider"), 60);
    });

    it("should open on hover", () => {
      const { getByTestId, getByRole } = render(
        <Slider min={0} max={100} data-testid="slider-root" />
      );

      const slider = getByRole("slider");
      const label = getByTestId("slider-root").querySelector(
        `.${sliderClassNames.thumb.label}`
      );
      // scale(0) makes it hidden
      expect(label).toHaveStyle({ transform: "translateY(-100%) scale(0)" });

      fireEvent.mouseOver(slider);

      expect(label).toBeVisible();
      expect(label).toHaveStyle({ transform: "translateY(-100%) scale(1)" });
    });

    it("should close on mouse leave", () => {
      const { getByTestId, getByRole } = render(
        <Slider min={0} max={100} data-testid="slider-root" />
      );

      const slider = getByRole("slider");
      const label = getByTestId("slider-root").querySelector(
        `.${sliderClassNames.thumb.label}`
      );
      expect(label).toHaveStyle({ transform: "translateY(-100%) scale(0)" });

      fireEvent.mouseOver(slider);

      expect(label).toHaveStyle({ transform: "translateY(-100%) scale(1)" });

      fireEvent.mouseLeave(slider);
      expect(label).toHaveStyle({ transform: "translateY(-100%) scale(0)" });
    });
  });

  describe("touch interaction", () => {
    it("should set value on touch", () => {
      const { getByTestId, getByRole } = render(
        <Slider min={0} max={100} data-testid="slider-root" />
      );

      const controlRoot = getControlRoot(getByTestId("slider-root"));
      fireEvent.touchStart(controlRoot, {
        changedTouches: [{ identifier: 0, clientX: 50 }],
      });

      expectSliderValue(getByRole("slider"), 50);
    });

    it("should set value on touch move", () => {
      const { getByTestId, getByRole } = render(
        <Slider min={0} max={100} data-testid="slider-root" />
      );

      const controlRoot = getControlRoot(getByTestId("slider-root"));
      fireEvent.touchStart(controlRoot, {
        changedTouches: [{ identifier: 0, clientX: 50 }],
      });
      fireEvent.touchMove(controlRoot, {
        changedTouches: [{ identifier: 0, clientX: 70 }],
      });

      expectSliderValue(getByRole("slider"), 70);
    });

    it("should ignore touch move from other identifier", () => {
      const { getByTestId, getByRole } = render(
        <Slider min={0} max={100} data-testid="slider-root" />
      );

      const controlRoot = getControlRoot(getByTestId("slider-root"));
      fireEvent.touchStart(controlRoot, {
        changedTouches: [{ identifier: 0, clientX: 50 }],
      });
      fireEvent.touchMove(controlRoot, {
        changedTouches: [{ identifier: 1, clientX: 70 }],
      });

      expectSliderValue(getByRole("slider"), 50);
    });

    it("should call onChange but no commit", () => {
      const onChange = vi.fn();
      const onChangeCommitted = vi.fn();
      const { getByTestId } = render(
        <Slider
          min={0}
          max={100}
          data-testid="slider-root"
          onChange={onChange}
          onChangeCommitted={onChangeCommitted}
        />
      );

      const controlRoot = getControlRoot(getByTestId("slider-root"));
      fireEvent.touchStart(controlRoot, {
        changedTouches: [{ identifier: 0, clientX: 50 }],
      });

      expect(onChange).toHaveBeenCalledWith({ value: 50 });
      expect(onChangeCommitted).not.toHaveBeenCalled();
    });

    it("should call onChangeCommitted", () => {
      const onChangeCommitted = vi.fn();
      const { getByTestId } = render(
        <Slider
          min={0}
          max={100}
          data-testid="slider-root"
          onChangeCommitted={onChangeCommitted}
        />
      );

      const controlRoot = getControlRoot(getByTestId("slider-root"));
      fireEvent.touchStart(controlRoot, {
        changedTouches: [{ identifier: 0, clientX: 50 }],
      });
      fireEvent.touchEnd(controlRoot, {
        changedTouches: [{ identifier: 0, clientX: 50 }],
      });

      expect(onChangeCommitted).toHaveBeenCalledWith({ value: 50 });
    });
  });

  describe("keyboard interaction", () => {
    it("should focus on tab", async () => {
      const { getByRole } = render(
        <Slider min={0} max={100} data-testid="slider-root" />
      );

      const slider = getByRole("slider");
      expect(document.activeElement).not.toEqual(slider);

      await userEvent.tab();

      expect(document.activeElement).toEqual(slider);
    });

    it("should move value", async () => {
      const { getByRole } = render(
        <Slider min={0} max={100} data-testid="slider-root" />
      );

      const slider = getByRole("slider");
      fireEvent.input(slider, { target: { valueAsNumber: 1 } });

      expectSliderValue(getByRole("slider"), 1);
    });

    it("should call onChange and commit", async () => {
      const onChange = vi.fn();
      const onChangeCommitted = vi.fn();
      const { getByRole } = render(
        <Slider
          min={0}
          max={100}
          onChange={onChange}
          onChangeCommitted={onChangeCommitted}
        />
      );

      const slider = getByRole("slider");
      fireEvent.input(slider, { target: { valueAsNumber: 1 } });
      expect(onChange).toHaveBeenCalledWith({ value: 1 });
      expect(onChangeCommitted).toHaveBeenCalledWith({ value: 1 });
    });

    it("should move value to next step", async () => {
      const { getByRole } = render(<Slider min={0} max={100} step={10} />);

      const slider = getByRole("slider");
      fireEvent.input(slider, { target: { valueAsNumber: 1 } });

      expectSliderValue(getByRole("slider"), 10);
    });

    it.each([
      [10, 11, 90],
      [90, 89, 10],
    ])(
      "should move value to next mark - from %s via %s to %s",
      async (from: number, ev: number, to: number) => {
        const { getByRole } = render(
          <Slider
            min={0}
            max={100}
            defaultValue={from}
            marks={[{ value: 10 }, { value: 90 }]}
            step="marks"
          />
        );

        const slider = getByRole("slider");
        fireEvent.input(slider, { target: { valueAsNumber: ev } });

        expectSliderValue(getByRole("slider"), to);
      }
    );

    it.each([
      [10, 9],
      [90, 91],
    ])(
      "should prevent moving out of bounds - from %s via %s",
      async (from: number, ev: number) => {
        const { getByRole } = render(
          <Slider
            min={0}
            max={100}
            defaultValue={from}
            marks={[{ value: 10 }, { value: 90 }]}
            step="marks"
          />
        );

        const slider = getByRole("slider");
        fireEvent.input(slider, { target: { valueAsNumber: ev } });

        expectSliderValue(getByRole("slider"), from);
      }
    );

    it("should snap to closest value if current value not snapped", () => {
      const { getByRole } = render(
        <Slider min={0} max={100} defaultValue={15} step={10} />
      );

      const slider = getByRole("slider");
      fireEvent.input(slider, { target: { valueAsNumber: 14 } });

      expectSliderValue(getByRole("slider"), 10);
    });
  });
});
