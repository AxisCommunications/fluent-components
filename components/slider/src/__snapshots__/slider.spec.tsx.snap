// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`slider > should render 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span
        class="axis-Slider ___1yr5lqm_13alxsp fp2oml8 fifp7yv fwrwuds f1asdtw4 f4aqd28 f22iagw f10pi13n fly5x3f f3er7iy f1lx44mx fke05om ftqa4ok f2hkw1w f1b1k54r f4ne723 fh7aioi fqqcjud f1k55ka9 fgclinu fycbxed f16pcs8n ffht0p2 f1p0ul1q f1c901ms f1alokd7 f78i1la f1kvsw7t f8k7e5g f1bw8brt f1pmxfrl fszkfcr f57dp0y f1ap5ily f11dm2qb f136rfnd fdsq1qd f1khssms fq4eyks f1ya6x16 f1e2iu44 ftuszwa f1pffoy2 f1bmyog6 fs6b7xr f15fv2gd"
      >
        <span
          class="axis-Slider__control ___144j0yg_julj5g0 f10pi13n f14t3ns0 f1l02sjl fcgxt0o f1ujusj6 fz0d3q9 f1k6fduh fdwl3z2"
        >
          <span
            class="axis-Slider__rail ___kotjr40_1djcq1g fnivh3a fc7yr5o f1el4m67 f8yange f1p9o1ba f1sil6mw ftgm304 f1euv43f fly5x3f f1erzq44 f1i1t8d1 f188r07x f1d3ncnn"
          >
            <span
              class="axis-Slider__track ___pq61zr0_1xft6bt fnivh3a fc7yr5o f1el4m67 f8yange ftgm304 f1euv43f f1erzq44 f1i1t8d1 f188r07x f1yrba31 f174n9yu"
              style="left: 0%; width: 20%;"
            />
          </span>
          <div
            class="axis-Slider__thumb ___14hb9vm_jmfl5w0 f8fbkgy f1nfllo7 f1djnp8u f1s8kh49 f22iagw f4d9j23 f122n59 f1euv43f f1t2h5zl f4aqd28 f1i1t8d1 f1vcxb5z fk1lali f1ewtqcl frcgkvu feh19e8 f3j3guj fo75joz f1yag6sl fsc6avr f1w15cjc fkaq55s f1fp5o0y f163fonl f1yq6w5o f11yjt3y f1jpmc5p f10tv6oz f16xp3sf fwrmqbx f1seuxxq f1j7ml58 f14u7mkt f5zrw40 fto0uou f1ks5ppg fyl8oag f1wl9k8s f1mdlcz9 fmmikit f10awi5f f1mfomsq fftmtmu f13zj6fq fgougqc fprarqb f14vs0nd f1gtfqs9 f18zvfd9 f8sy9bk fufffdp f1sclg23 f132a78j fcetbih f1kt6s4o"
            style="left: 20%;"
          >
            <input
              data-index="0"
              id="slider-thumb-1"
              max="100"
              min="0"
              type="range"
              value="20"
            />
            <span
              class="axis-Slider__thumb-label ___1b1bz8r_1eqf7oo fk116j8 fg8tfde fpvge99 fv71qf3 f1ywm7hm fw5db7e f14wxoun f1uw59to f1aa9q02 f16jpd5f f1jar5jt fyu767a f1euv43f feml9b3 f22iagw f122n59 f4d9j23 fxugw4r f18ygdci fxeb0a7 f1j7ml58 f1wl9k8s f15w9qf9 f1js41m7 fguqkdo f1ljr5q2 fto0uou f1hr5hl2 f1brpl1h"
            >
              20
            </span>
          </div>
        </span>
      </span>
    </div>
  </body>,
  "container": <div>
    <span
      class="axis-Slider ___1yr5lqm_13alxsp fp2oml8 fifp7yv fwrwuds f1asdtw4 f4aqd28 f22iagw f10pi13n fly5x3f f3er7iy f1lx44mx fke05om ftqa4ok f2hkw1w f1b1k54r f4ne723 fh7aioi fqqcjud f1k55ka9 fgclinu fycbxed f16pcs8n ffht0p2 f1p0ul1q f1c901ms f1alokd7 f78i1la f1kvsw7t f8k7e5g f1bw8brt f1pmxfrl fszkfcr f57dp0y f1ap5ily f11dm2qb f136rfnd fdsq1qd f1khssms fq4eyks f1ya6x16 f1e2iu44 ftuszwa f1pffoy2 f1bmyog6 fs6b7xr f15fv2gd"
    >
      <span
        class="axis-Slider__control ___144j0yg_julj5g0 f10pi13n f14t3ns0 f1l02sjl fcgxt0o f1ujusj6 fz0d3q9 f1k6fduh fdwl3z2"
      >
        <span
          class="axis-Slider__rail ___kotjr40_1djcq1g fnivh3a fc7yr5o f1el4m67 f8yange f1p9o1ba f1sil6mw ftgm304 f1euv43f fly5x3f f1erzq44 f1i1t8d1 f188r07x f1d3ncnn"
        >
          <span
            class="axis-Slider__track ___pq61zr0_1xft6bt fnivh3a fc7yr5o f1el4m67 f8yange ftgm304 f1euv43f f1erzq44 f1i1t8d1 f188r07x f1yrba31 f174n9yu"
            style="left: 0%; width: 20%;"
          />
        </span>
        <div
          class="axis-Slider__thumb ___14hb9vm_jmfl5w0 f8fbkgy f1nfllo7 f1djnp8u f1s8kh49 f22iagw f4d9j23 f122n59 f1euv43f f1t2h5zl f4aqd28 f1i1t8d1 f1vcxb5z fk1lali f1ewtqcl frcgkvu feh19e8 f3j3guj fo75joz f1yag6sl fsc6avr f1w15cjc fkaq55s f1fp5o0y f163fonl f1yq6w5o f11yjt3y f1jpmc5p f10tv6oz f16xp3sf fwrmqbx f1seuxxq f1j7ml58 f14u7mkt f5zrw40 fto0uou f1ks5ppg fyl8oag f1wl9k8s f1mdlcz9 fmmikit f10awi5f f1mfomsq fftmtmu f13zj6fq fgougqc fprarqb f14vs0nd f1gtfqs9 f18zvfd9 f8sy9bk fufffdp f1sclg23 f132a78j fcetbih f1kt6s4o"
          style="left: 20%;"
        >
          <input
            data-index="0"
            id="slider-thumb-1"
            max="100"
            min="0"
            type="range"
            value="20"
          />
          <span
            class="axis-Slider__thumb-label ___1b1bz8r_1eqf7oo fk116j8 fg8tfde fpvge99 fv71qf3 f1ywm7hm fw5db7e f14wxoun f1uw59to f1aa9q02 f16jpd5f f1jar5jt fyu767a f1euv43f feml9b3 f22iagw f122n59 f4d9j23 fxugw4r f18ygdci fxeb0a7 f1j7ml58 f1wl9k8s f15w9qf9 f1js41m7 fguqkdo f1ljr5q2 fto0uou f1hr5hl2 f1brpl1h"
          >
            20
          </span>
        </div>
      </span>
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
